nova_vm/src/ecmascript/builtins/structured_data/data_view_objects/data_view_constructor.rs

```diff
@@ -24,14 +34,123 @@ impl BuiltinIntrinsicConstructor for DataViewConstructor {
+        arguments: ArgumentsList,
+        new_target: Option<Object>,
     ) -> JsResult<Value> {
-        todo!()
+        // 1. If NewTarget is undefined, throw a TypeError exception.
+        let Some(new_target) = new_target else {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "class constructors must be invoked with 'new'",
+            ));
+        };
+        let new_target = Function::try_from(new_target).unwrap();
+
+        let buffer = arguments.get(0);
+        let byte_offset = arguments.get(1);
+        let byte_length = arguments.get(2);
+
+        // 2. Perform ? RequireInternalSlot(buffer, [[ArrayBufferData]]).
+        let buffer = require_internal_slot_array_buffer(agent, buffer)?;
+
+        // 3. Let offset be ? ToIndex(byteOffset).
+        let offset = to_index(agent, byte_offset)? as usize;
+
+        // 4. If IsDetachedBuffer(buffer) is true, throw a TypeError exception.
+        if is_detached_buffer(agent, buffer) {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "attempting to access detached ArrayBuffer",
+            ));
+        }
+
+        // 5. Let bufferByteLength be ArrayBufferByteLength(buffer, seq-cst).
+        let buffer_byte_length = array_buffer_byte_length(agent, buffer, Ordering::SeqCst) as usize;
+
+        // 6. If offset > bufferByteLength, throw a RangeError exception.
+        if offset > buffer_byte_length {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::RangeError,
+                "offset is outside the bounds of the buffer",
+            ));
+        }
+
+        // 7. Let bufferIsFixedLength be IsFixedLengthArrayBuffer(buffer).
+        let buffer_is_fixed_length = is_fixed_length_array_buffer(agent, buffer);
+
+        // 8. If byteLength is undefined, then
+        let view_byte_length = if byte_length.is_undefined() {
+            // a. If bufferIsFixedLength is true, then
+            if buffer_is_fixed_length {
+                // i. Let viewByteLength be bufferByteLength - offset.
+                Some(buffer_byte_length - offset)
+            } else {
+                // b. Else,
+                // i. Let viewByteLength be auto.
+                None
+            }
+        } else {
+            // 9. Else,
+            // a. Let viewByteLength be ? ToIndex(byteLength).
+            let view_byte_length = to_index(agent, byte_length)? as usize;
+            // b. If offset + viewByteLength > bufferByteLength, throw a RangeError exception.
+            if offset + view_byte_length > buffer_byte_length {
+                return Err(agent.throw_exception_with_static_message(
+                    ExceptionType::RangeError,
+                    "offset is outside the bounds of the buffer",
+                ));
+            }
+            Some(view_byte_length)
```

Comment:
**note:** If we end up deciding to make `AUTO` the "default" value, then we should here also conditionally return `None` for fixed length ABs that have `offset + view_byte_length == buffer_byte_length`.